File: vripper-server/src/main/java/tn/mnlr/vripper/services/XpathService.java
package tn.mnlr.vripper.services;

import org.springframework.stereotype.Service;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import tn.mnlr.vripper.exception.XpathException;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;

@Service
public class XpathService {

    private final XPath xpath = XPathFactory.newInstance().newXPath();

    public Node getAsNode(Node source, String xpathExpression) throws XpathException {
        try {
            return (Node) xpath.compile(xpathExpression).evaluate(source, XPathConstants.NODE);
        } catch (Exception e) {
            throw new XpathException(e);
        }
    }

    public NodeList getAsNodeList(Node source, String xpathExpression) throws XpathException {
        try {
            return (NodeList) xpath.compile(xpathExpression).evaluate(source, XPathConstants.NODESET);
        } catch (Exception e) {
            throw new XpathException(e);
        }
    }

}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/services/XpathService.kt
package me.mnlr.vripper.services

import org.springframework.stereotype.Service
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import me.mnlr.vripper.exception.XpathException
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathFactory

@Service
class XpathService {
    private val xpath = XPathFactory.newInstance().newXPath()

    @Throws(XpathException::class)
    fun getAsNode(source: Node?, xpathExpression: String?): Node? {
        return try {
            xpath.compile(xpathExpression).evaluate(source, XPathConstants.NODE) as Node
        } catch (e: Exception) {
            throw XpathException(e)
        }
    }

    @Throws(XpathException::class)
    fun getAsNodeList(source: Node?, xpathExpression: String?): NodeList {
        return try {
            xpath.compile(xpathExpression).evaluate(source, XPathConstants.NODESET) as NodeList
        } catch (e: Exception) {
            throw XpathException(e)
        }
    }
}